roachtest: psycopg failed

**Note:** This build has runtime assertions enabled. If the same failure was hit in a run without assertions enabled, there should be a similar failure without this message. If there isn&#39;t one, then this failure is likely due to an assertion violation or (assertion) timeout.

roachtest.psycopg [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/19712894?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/19712894?buildTab=artifacts#/psycopg) on release-25.2 @ [78cb231bf4a068dbafa877d2691c7042b8558191](https://github.com/cockroachdb/cockroach/commits/78cb231bf4a068dbafa877d2691c7042b8558191):


```
(orm_helpers.go:224).summarizeFailed: 
Tests run on Cockroach v25.2.0-rc.1-dev-78cb231bf4a068dbafa877d2691c7042b8558191
Tests run against psycopg 3.2.8
3928 Total Tests Run
3927 tests passed
1 test failed
1919 tests skipped
30 tests ignored
0 tests passed unexpectedly
1 test failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: tests.pool.test_pool.test_shrink - unknown (unexpected)
For a full summary look at the psycopg artifacts 
An updated blocklist (psycopgBlockList) is available in the artifacts' psycopg log
test artifacts and logs in: /artifacts/psycopg/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>metamorphicBufferedSender=true</code>
 - <code>metamorphicLeases=expiration</code>
 - <code>runtimeAssertionsBuild=true</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-19712894/psycopg/1747466564844/1747467433532)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #146427 roachtest: psycopg failed [C-test-failure O-roachtest O-robot P-3 T-sql-foundations branch-release-25.2.0-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*psycopg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

```
dsn = 'host=localhost port=26257 user=roachprod password=cockroachdb dbname=defaultdb'
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x757b9c50b250>

    @pytest.mark.slow
    @pytest.mark.timing
    def test_shrink(dsn, monkeypatch):
        from psycopg_pool.pool import ShrinkPool
    
        results: list[tuple[int, int]] = []
    
        def run_hacked(self, pool):
            n0 = pool._nconns
            orig_run(self, pool)
            n1 = pool._nconns
            results.append((n0, n1))
    
        orig_run = ShrinkPool._run
        monkeypatch.setattr(ShrinkPool, "_run", run_hacked)
    
        def worker(n):
            with p.connection() as conn:
                conn.execute("select pg_sleep(0.1)")
    
        with pool.ConnectionPool(dsn, min_size=2, max_size=4, max_idle=0.2) as p:
            p.wait(5.0)
            assert p.max_idle == 0.2
    
            ts = [spawn(worker, args=(i,)) for i in range(4)]
            gather(*ts)
    
            sleep(1)
    
>       assert results == [(4, 4), (4, 3), (3, 2), (2, 2), (2, 2)]
E       assert [(4, 4), (4, 3), (3, 2), (2, 2), (2, 2), (2, 2)] == [(4, 4), (4, 3), (3, 2), (2, 2), (2, 2)]
E         
E         Left contains one more item: (2, 2)
E         
E         Full diff:
E           [
E               (
E                   4,
E                   4,
E               ),
E               (
E                   4,
E                   3,
E               ),
E               (
E                   3,
E                   2,
E               ),
E               (
E                   2,
E                   2,
E               ),
E               (
E                   2,
E                   2,
E               ),
E         +     (
E         +         2,
E         +         2,
E         +     ),
E           ]

tests/pool/test_pool.py:476: AssertionError
```

This may be a flaky psycopg test.

--- a/pkg/cmd/roachtest/tests/psycopg_blocklist.go
+++ b/pkg/cmd/roachtest/tests/psycopg_blocklist.go
@@ -24,6 +24,7 @@ var psycopgIgnoreList = blocklist{
 	`tests.pool.test_pool.test_reconnect_after_grow_failed`:                                          "requires insecure mode",
 	`tests.pool.test_pool.test_reconnect_failure[False]`:                                             "requires insecure mode",
 	`tests.pool.test_pool.test_refill_on_check`:                                                      "requires insecure mode",
+	`tests.pool.test_pool.test_shrink`:                                                               "flaky; see #146895",
 	`tests.pool.test_pool.test_stats_connect`:                                                        "requires insecure mode",
 	`tests.pool.test_pool_async.test_connect_check_timeout[asyncio]`:                                 "requires insecure mode",
 	`tests.pool.test_pool_async.test_reconnect[asyncio]`:                                             "requires insecure mode",

